draft-ietf-masque-h3-datagram.md

```diff
@@ -195,41 +195,93 @@ silently or buffer it temporarily while awaiting the creation of the
+bidirectional stream of bytes that follow the headers in both directions. In
+HTTP/1.x, the data stream consists of all bytes on the connection that follow
+the blank line that concludes the header section in both directions. In HTTP/2
+and HTTP/3, the data stream of a given HTTP request consists of all bytes sent
+in DATA frames with the corresponding stream ID.
+
+Definitions of new HTTP Methods or of new HTTP Upgrade Tokens can state that
+their data stream uses the Capsule Protocol. If they do so, that means that the
+contents of their data stream uses the following format (using the notation
+from the "Notational Conventions" section of {{QUIC}}):
 
 ~~~
-CAPSULE HTTP/3 Frame {
-  Type (i) = 0xffcab5,
-  Length (i),
-  Capsule Type (i),
-  Capsule Data (..),
+Capsule Protocol {
+  Capsule (..) ...,
 }
 ~~~
-{: #capsule-frame-format title="CAPSULE HTTP/3 Frame Format"}
+{: #capsule-stream-format title="Capsule Protocol Stream Format"}
 
-The Type and Length fields follows the definition of HTTP/3 frames from {{H3}}.
-The payload consists of:
+~~~
+Capsule {
+  Capsule Type (i),
+  Capsule Length (i),
+  Capsule Value (..),
+}
+~~~
+{: #capsule-format title="Capsule Format"}
 
 Capsule Type:
 
-: The type of this capsule.
+: A variable-length integer indicating the Type of the capsule. Endpoints that
+receive a capsule with an unknown Capsule Type MUST silently skip over that
+capsule.
+
+Capsule Length:
+
+: The length of the Capsule Value field following this field, encoded as a
+variable-length integer. Note that this field can have a value of zero.
+
+Capsule Value:
+
+: The payload of this capsule. Its semantics are determined by the value of the
+Capsule Type field.
+
+
+## Requirements
+
+If the definition of an HTTP Method or HTTP Upgrade Token states that it uses
+the capsule protocol, its implementations MUST follow the following
+requirements:
+
+* A server MUST NOT send any Transfer-Encoding or Content-Length header fields
+  in a 2xx (Successful) response. If a client receives a Content-Length or
+  Transfer-Encoding header fields in a successful response, it MUST treat that
+  response as malformed.
+
+* A request message does not have content.
+
+* Responses are not cacheable.
 
-Capsule Data:
+## Intermediary Processing
 
-: Data whose semantics depends on the Capsule Type.
+Intermediaries MUST operate in one of the two following modes:
+
+Pass-through mode:
+
+: In this mode, the intermediary forwards the data stream between two
+associated streams without any modification of the data stream.
+
+Participant mode:
+
+: In this mode, the intermediary terminates the data stream and parses all
+Capsule Type and Capsule Length fields it receives.
+
+Each Capsule Type determines whether it is opaque or transparent to
+intermediaries in participant mode: opaque capsules are forwarded unmodified
+while transparent ones can be parsed, added, or removed by intermediaries.
```

Comment:
assume "modification" of a capsule is really remove and add?

Reply:
Clarified that modification of the data of transparent capsules is OK, after all it's isomorphic to remove + add